Add docs about Scheduler HA, how to use it and DB requirements (#11467)
mistercrunch/Airflow: docs/scheduler.rst

This only has effect if your DAG has no ``schedule_interval``.
If you keep default ``allow_trigger_in_future = False`` and try 'external trigger' to run future-dated execution dates,
the scheduler won't execute it now but the scheduler will execute it in the future once the current date rolls over to the execution date.

> Running more than one scheduler

Reviewer: I would suggest to use sentence case to keep consistent with other titles in this page

-------------------------------

> Airflow 2.0 is the first release that officially supports running more one scheduler concurrently -- both for

Reviewer: `... more one scheduler...` -> `... more than one scheduler...`

> Airflow 2.0 is the first release that officially supports running more one scheduler concurrently -- both for

Reviewer 2: ```suggestion
Airflow supports running more than one scheduler concurrently -- both for
```
Shorter and will be longer "up to date"

performance reasons and for resiliency.

Overview
""""""""

> The HA scheduler is written to take advantage of the existing metadata database. This was primarily done for

Reviewer 2: ```suggestion
The highly available (HA) scheduler is designed to take advantage of the existing metadata database. This was primarily done for
```
Not every user have to be familiar with this acronym

operational simplicity: every component already has to speak to this DB, and by not using direct communication
or consensus algorithm between schedulers (Raft, Paxos, etc.) nor another consensus tool (Apache Zookeeper, or

> Consul for instance) we have kept the "operational surface area" to a minimum.

Reviewer 2: ```suggestion
Consul for instance) we have kept the "operational surface area" to a minimum. For more information on architectural choices check `Airflow Improvement Proposal 15 <https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=103092651>`_.
```

Author: Hmmm, not sure about this one -- I like the idea, but I think I'd rather the docs be the source of truth rather than linking to the AIP.

I might link to the AIP for now though.

> The scheduler now uses the serialized DAG representation to make it's scheduling decisions and the rough

Reviewer: `it's` -> `its`

outline of the scheduling loop is:

- Check for any DAGs needing a new DagRun, and create them
- Examine a batch of DagRuns for schedulable TaskInstances or complete DagRuns
- Select schedulable TaskInstances, and whilst respecting Pool limits and other concurrency limits, enqueue
  them for execution

This does however place some requirements on the Database.

Database Requirements
"""""""""""""""""""""

> The short version is that users of PostgreSQL 9.6+ or MySQL 8+ are all ready to go -- simple start running as

Reviewer: `simple start...` -> `simply start...`

> The short version is that users of PostgreSQL 9.6+ or MySQL 8+ are all ready to go -- simple start running as

Author: "simply" and "just" are bad words to use in cos `simple` -> `you can`

many copies of the scheduler as you like -- there is no further set up or config options needed. If you are

> using a different database please read on

Reviewer 3: ```suggestion
using a different database please read on.
```

To maintain performance and throughput there is one part of the scheduling loop that does a number of
calculations in memory (because having to round-trip to the DB for each TaskInstance would be too slow) so we
need to ensure that only a single scheduler is in this critical section at once - otherwise limits would not
be correctly respected. To achieve this we use database row-level locks (using ``SELECT ... FOR UPDATE``).

This critical section is where TaskInstances go from scheduled state and are enqueued to the executor, whilst
ensuring the various concurrency and pool limits are respected. The critical section is obtained by asking for
a row-level write lock on every row of the Pool table (roughly equivalent to ``SELECT * FROM slot_pool FOR
UPDATE NOWAIT`` but the exact query is slightly different).

The following databases are fully supported and provide an "optimal" experience:

- PostgreSQL 9.6+
- MySQL 8+

> MaraiDB will have slightly "degraded" performance, as it does not implement the ``SKIP LOCKED`` or ``NOWAIT`` SQL clauses --

Reviewer 4: On line 115, "MaraiDB" should be changed to "MariaDB".

Reviewer 5: Hey @vikramkoka -> FYI. there is this nice feature "suggest change" which is really cool for those kind of typos. It's not easily discoverable - I initially also did not know that it exist, but once I started using it, I even change small snippets of code this way :)

![Screenshot from 2020-10-13 15-24-57](https://user-images.githubusercontent.com/595491/95866655-783dad80-0d68-11eb-855d-4ee148a10362.png)

> MaraiDB will have slightly "degraded" performance, as it does not implement the ``SKIP LOCKED`` or ``NOWAIT`` SQL clauses --

Reviewer 6: Maybe worth clarifying MariaDB can only be ran with one scheduler currently here (per https://apache-airflow.slack.com/archives/CCVDRN0F9/p1603564032117100?thread_ts=1603544070.099000&cid=CCVDRN0F9)

Reviewer 5: I think we should also update information about supported versions in https://github.com/apache/airflow#requirements - with some * disclaimer that the HA Scheduler works in MySQL 8 but both Maria DB and Mysql 5.7 have limitations.

Author: I think this may a limit specific to Galera cluster, not MariaDB itself.

see `MDEV-13115 <https://jira.mariadb.org/browse/MDEV-13115>`_. Multiple schedulers will operate without this
extension, but instead of skipping over the critical section and carrying on with other scheduling work
(creating DagRuns, progressing TaskInstances to the scheduled state) they will wait for the critical section
lock to be released.

.. warning::

  MySQL 5.x also does not support ``SKIP LOCKED`` or ``NOWAIT``, and additionally is more prone to deciding

>   queries are deadlocked, so running with a more than a single scheduler on MySQL 5.x is not supported or

Reviewer: `...running with a more than a single scheduler...` -> `...running with more than a single scheduler...`

  recommended.

.. note::

  Microsoft SQLServer has not been tested with HA.

Tuneables
"""""""""

> - :ref:`config:scheduler__max_dagruns_to_create_per_loop`

Author: I'm almost tempted to move the docs out of the config and include them here, or expand on them here and just leave a short version in the config and point at this section from there.

Reviewer 3: I will vote for: Expanding them here and just leave a short version in the config and point at this section from there.

Reviewer: I agree with Kaxil

Author: Done

  queries are deadlocked, so running with more than a single scheduler on MySQL 5.x is not supported or
  recommended.

.. note::

  Microsoft SQLServer has not been tested with HA.

.. _scheduler:ha:tunables:

Scheduler Tuneables
"""""""""""""""""""

The following config settings can be used to control aspects of the Scheduler HA loop.

- :ref:`config:scheduler__max_dagruns_to_create_per_loop`

  This changes the number of dags that are locked by each scheduler when
  creating dag runs. One possible reason for setting this lower is if you
  have huge dags and are running multiple schedules, you won't want one
  scheduler to do all the work.

- :ref:`config:scheduler__max_dagruns_per_loop_to_schedule`

  How many DagRuns should a scheduler examine (and lock) when scheduling
  and queuing tasks. Increasing this limit will allow more throughput for
  smaller DAGs but will likely slow down throughput for larger (>500
  tasks for example) DAGs. Setting this too high when using multiple
  schedulers could also lead to one scheduler taking all the dag runs
  leaving no work for the others.

- :ref:`config:scheduler__use_row_level_locking`

>   Should the scheduler issue `SELECT ... FOR UPDATE` in relevant queries.
>   If this is set to False then you should not run more than a single
>   scheduler at once

Reviewer 3: ```suggestion
  Should the scheduler issue ``SELECT ... FOR UPDATE`` in relevant queries.
  If this is set to ``False`` then you should not run more than a single
  scheduler at once.
```
